### PR TITLE
feat(core): grantErrorListener for logging token exchange error

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

Add grantErrorListener for logging token exchange error

## Linear Issue Reference

LOG-325
LOG-2421

## Testing

image

### CODE REVIEW

File: packages/core/src/utils/oidc-provider-event-listener.ts [hunk 1]
```diff
@@ -49,3 +49,27 @@ export const grantSuccessListener = async (
...
+      params,
+    },
+  } = ctx;
+  ctx.addLogContext({
+    applicationId: client?.clientId,
+  });
+
+  const grantType = params?.grant_type;
+  const type: LogType =
+    grantType === GrantType.AuthorizationCode ? 'CodeExchangeToken' : 'RefreshTokenExchangeToken';
```
Comment (demonzoo) [inline packages/core/src/utils/oidc-provider-event-listener.ts:69]: You can consider extracting this `getType` logic into a static method

**Review by demonzoo — APPROVED** [review]
Overall looks good

File: packages/core/src/utils/oidc-provider-event-listener.ts (hunk 1 above)
Comment (IceHe) [reply packages/core/src/utils/oidc-provider-event-listener.ts]: Refactored. PTAL.

File: packages/core/src/utils/oidc-provider-event-listener.ts [hunk 2]
```diff
@@ -39,13 +42,35 @@ export const grantSuccessListener = async (
...
+      entities: { Client: client },
+      params,
+    },
+  } = ctx;
+  ctx.addLogContext({
+    applicationId: client?.clientId,
+  });
+
+  const grantType: unknown = params?.grant_type;
+  const type = getLogType(grantType as GrantType);
```
Comment (demonzoo) [inline packages/core/src/utils/oidc-provider-event-listener.ts:70]: Nope, as grantType can be `undefined`, so it at least should be `Optional<GrantType>`

Can you try something like this:
```typescript
const type = getLogType(params?.grant_type as Optional<GrantType>);
```

And in the static method, define grantType as an optional param
```typescript
const getLogType = (grantType?: GrantType): LogType => ...

Comment (IceHe) [reply packages/core/src/utils/oidc-provider-event-listener.ts]: Refactored. PTAL

Comment (demonzoo) [reply packages/core/src/utils/oidc-provider-event-listener.ts]: nit: Add a colon to the end of the error message

Comment (IceHe) [reply packages/core/src/utils/oidc-provider-event-listener.ts]: Updated and add new UTs of skipping logging. PTAL

```ts
    console.error('Unexpected grant_type:', grantType);
```

**Pull request merged**